Refactor receiver webhook client to support custom dialers

JacobsonMT opened the pull request:
- Introduced client options in HTTP client creation similar to upstream.
- Added support for custom `DialContext` in webhook client.
- Enhanced integration tests to validate custom dialer behavior.

Part of a series of PRs for full support. This PR lacks support for: mqtt, sns, slack, and prometheus-alertmanager. It also leaves more in-depth client refactoring to followups.

Most changes to tests are to facilitate the new `factory_test` that verifies a provided custom dialer blocking requests works as expected. Various issues with the "valid" testing configs for integrations were causing them to fail for other reasons before they hit the dialer.

JacobsonMT committed:
Refactor receiver webhook client to support custom dialers

- Introduced client options in HTTP client creation similar to upstream.
- Added support for custom `DialContext` in webhook client.
- Enhanced integration tests to validate custom dialer behavior.

Part of a series of PRs for full support. This PR lacks support
for: mqtt, sns, slack, and prometheus-alertmanager. It also
leaves more in-depth client refactoring to followups.

JacobsonMT committed:
Linting

rwwiv commented on http/client.go:
```diff
@@ -18,23 +20,53 @@ import (
+			}),
+		}
+	}
+)
```
We can just pull this into the `NewClient` constructor now.

rwwiv commented on http/client.go:
```diff
@@ -18,23 +20,53 @@ import (
+
+func WithDialer(dialer net.Dialer) ClientOption {
+	return func(c *clientConfiguration) {
+		c.dialer = dialer
```
Any reason we're using a mutator pattern here instead of an options struct?

JacobsonMT replied:
Mostly because upstream uses a mutator pattern in its equivalent implementation. Figured there was some small benefit to keeping the implementations similar.

Other than that:
- It makes the usage pattern fairly clear. For example, that the configuration is optional, I have a slight preference towards this over passing in `nil` or exporting a `DefaultClientConfiguration` struct.
- Allows us to more easily provide selective modifications to the client per integration when necessary. For example, letting just `Jira` perform `GET` requests, or if we move `TLS`/`HMAC` config to `ClientOption` it can allows us to gradually implement support for those in other receivers without needing a `ClientConfiguration` "clone"/"apply option" flow.

These are not super important though, if you have a strong preference.

rwwiv replied:
To your points:
- You can abuse variadic parameters to avoid needing to pass in a `nil` or default option (e.g. `... Option`) 
- I'm not sure I follow here, either method lets us do this no? A new field on the option struct would accomplish the same thing that another mutator would.


Here's an example of how this would look:

```go
type Client struct {
    log logging.Logger
    cfg clientConfiguration
}

type clientConfiguration struct {
    userAgent string
    dialer    *net.Dialer
}

type Options struct {
    UserAgent string
    Dialer    *net.Dialer
}

func NewClient(log logging.Logger, opts ...Options) *Client {
    // Default configuration
    cfg := clientConfiguration{
        userAgent: "Grafana",
        dialer: &net.Dialer{
            Timeout: 30 * time.Second,
        },
    }

    // Apply options if provided
    if len(opts) > 0 {
        opt := opts[0]
        if opt.UserAgent != "" {
            cfg.userAgent = opt.UserAgent
        }
        if opt.Dialer != nil {
            cfg.dialer = opt.Dialer
        }
    }

    return &Client{
        log: log,
        cfg: cfg,
    }
}

func (c *Client) Dialer() *net.Dialer {
    return c.cfg.dialer
}
```

JacobsonMT replied:
Thanks for the example, that’s helpful.

I see what you mean, and you're right that both patterns support selective configuration in the end. Personally, I tend to dislike using variadic parameters to simulate optionality, since it can feel like an abuse of the language and sometimes obscure usage expectations for callers.

That said, I don't think there's a huge technical advantage either way. It's mostly a stylistic choice, and I’m happy to go with the options struct if you have a strong preference. Just let me know.

JacobsonMT committed:
Inline defaultClientConfiguration

yuri-tceretian commented on http/client.go:
```diff
@@ -18,23 +20,47 @@ import (
+		}),
+	}
+	for _, opt := range opts {
+		opt(&cfg)
```
I think we should be a little more defensive and check for `opt` to be nil here

yuri-tceretian commented on http/client.go:
```diff
@@ -18,23 +20,47 @@ import (
+}
+
+func (ns *Client) Dialer() *net.Dialer {
+	return &ns.cfg.dialer
```
Can we return copy of the dialer instead? Returning pointer allows client to be implicitly mutated.

yuri-tceretian commented on http/client.go:
```diff
@@ -18,23 +20,47 @@ import (
+	cfg := clientConfiguration{
+		userAgent: "Grafana",
+		dialer: (net.Dialer{
+			Timeout: 30 * time.Second,
```
parenthesis seem to be unnecessary. Where does this 30s come from?

JacobsonMT replied:
Not sure, possibly from the `DefaultTransport` in stlib? Mostly just trying to keep unrelated functional changes to a minimum, so kept the `30s` default. Our client timeout is also `30s` which should cover ensuring the dial timeout remains bounded even when a custom dialer is provided.

yuri-tceretian replied:
ok, let's annotate those constants with comments and mention the reason why they are there

JacobsonMT replied:
> Our client timeout is also 30s which should cover ensuring the dial timeout remains bounded even when a custom dialer is provided.

On second thought, I'm no longer sure this makes sense. So, I've added a fallback to ensure the dial timeout is configured if given a custom dialer without one set. Maybe it's not necessary, but it won't hurt.

Added some comments as well.

yuri-tceretian commented on http/tls.go:
```diff
@@ -1,31 +1,34 @@
-	return nc(tlsConfig)
+	return &http.Client{
+		Timeout: time.Second * 30,
+		Transport: &http.Transport{
```
Perhaps we should set IdleConnTimeout here?

JacobsonMT replied:
Similar to above, I agree our client settings need improvement but I don't want to change them in this PR.

yuri-tceretian commented on http/tls.go:
```diff
@@ -1,31 +1,34 @@
 
-	return nc(tlsConfig)
+	return &http.Client{
+		Timeout: time.Second * 30,
```
This way we basically cap all downstream timeouts handshake, sending request, receiving headers and body  etc, including dialer. I think this could lead to some hidden problems. Why do we set it to 30s? And if we do need it, probably we should control other timeouts too and do some coordination.

JacobsonMT replied:
Similar to above, I agree our client settings need improvement but I don't want to change them in this PR.

yuri-tceretian commented:
I am not sure I understand where the hard-coded timeouts came from. Also, I am worried about coordination of hard-coded and configured timeouts. Other than that, LGTM

JacobsonMT committed:
Address PR comments

yuri-tceretian reviewed (approved):
LGTM

santihernandezc commented on http/client.go:
```diff
@@ -18,23 +20,49 @@ import (
+
+func (ns *Client) Dialer() net.Dialer {
+	return ns.cfg.dialer
+}
```
Is there a plan to use this method in other repos? I couldn't find any calls here.

JacobsonMT replied:
It's used in the followup PR. Accidentally left it in this one.

santihernandezc commented on notify/factory_test.go:
```diff
@@ -72,6 +76,56 @@ func TestBuildReceiverIntegrations(t *testing.T) {
+				http.WithUserAgent("Grafana-test"),
+				http.WithDialer(net.Dialer{
+					// Also override the Resolver so that configuration that doesn't use real hostnamesalso return
+					// "custom dial function error" instead of "no such dns".
```
Super nit, there's a space missing.
```suggestion
					// Also override the Resolver so that configuration that doesn't use real hostnames
					// also return "custom dial function error" instead of "no such dns".
```

santihernandezc reviewed (approved):
LGTM!
I agree with Yuri's comments about timeouts, but I understand this PR might not be the best place to fix it. I created an [issue to make timeouts configurable](https://github.com/grafana/alerting/issues/324).

JacobsonMT committed:
Refactor `BuildReceiverIntegrations` to use variadic `httpClientOptions`

JacobsonMT committed:
Remove unused Dialer method

JacobsonMT committed:
Improve comment clarity in factory_test.go regarding Resolver override behavior

JacobsonMT committed:
Ensure dialer timeout is set when given custom dialer

Falls back to previous default timeout of 30s, this is mostly
done defensively to minimize functional differences.
